### ISSUE TITLE
Many packages does not distinguish halfwidth and fullwidth characters

### ISSUE THREAD

**Issue by simnalamburt** [remarkjs/remark-lint#310]
### Initial checklist

- [X] I read the [support docs](https://github.com/remarkjs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/remarkjs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/remarkjs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Aremarkjs&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

remark-lint-table-pipe-alignment@3.1.3

### Steps to reproduce

Open below with browser:

```html
<!doctype html>
<script type="module">
import {reporter} from 'https://esm.sh/vfile-reporter@8.1.0?bundle'
import {remark} from 'https://esm.sh/remark@15.0.1?bundle'
import remarkGfm from 'https://esm.sh/remark-gfm@4.0.0?bundle'
import remarkLint from 'https://esm.sh/remark-lint@9.1.2?build'
import remarkLintTablePipeAlignment from 'https://esm.sh/remark-lint-table-pipe-alignment@3.1.3?bundle'

const file = await remark()
  .use(remarkGfm)
  .use(remarkLint)
  .use(remarkLintTablePipeAlignment)
  .process(`
False positive
--------
Below should be OK but linter will report it as error.

| Alpha | 한글 |
|-------|------|
| a     | b    |

False negative
--------
Below should be error but linter will report it as OK.

| Alpha | 한글 |
|-------|------|
| a     | b  |
`)

document.body.innerHTML = `<pre>${reporter(file)}</pre>`
</script>

```

### Expected behavior

The first one should be fine and the second one should be an error.

```
16:16-16:17 warning Misaligned table fence table-pipe-alignment remark-lint

⚠ 1 warning
```

### Actual behavior

However, the first one gives an error and the second one does not.

```
8:16-8:17 warning Misaligned table fence table-pipe-alignment remark-lint

⚠ 1 warning
```

### Runtime

Node v17, Node v16, Node v14, Node v12, Deno, Electron, Other (please specify in steps to reproduce)

### Package manager

npm 8, npm 7, npm 6, yarn 2, yarn 1, pnpm, Other (please specify in steps to reproduce)

### OS

Windows, Linux, macOS, Other (please specify in steps to reproduce)

### Build and bundle tools

Webpack, Rollup, esbuild, Parcel, Create React App, Gatsby, Next.js, Remix, Docusaurus, Snowpack, Vite, Other (please specify in steps to reproduce)

**Comment by wooorm** [comment]
Hey!

This rule measures characters.
The size of displayed characters is different in different places. Characters will be different in a terminal, different here on GitHub, different in your editor, different in my editor.

There isn’t a good solution. I recommend turning this rule off instead. See https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-table-pipe-alignment#fix for more info

**Comment by simnalamburt** [comment]
Majority of programmers codes with fixed sized font and width of CJK characters are well defined and documented in here: https://unicode.org/Public/UNIDATA/EastAsianWidth.txt

We at least need to create a new option to respect the size of full width character. Something like "fixed width font mode"

**Comment by wooorm** [comment]
Have you tested your ideas in different code editors?

**Comment by simnalamburt** [comment]
I'm Korean. Yes I have been used CJK letters in many editors including vi, nvim, Vim, nano, micro, ed, terminal, emacs, VSCode, atom, zed, notepad.exe, notepad++, Visual Studio, sublime. In these editors, CJK letters will be displayed as fullwidth character (which have double width of halfwidth character) unless you explicitly change your font into something other than fixed width font.

**Comment by simnalamburt** [comment]
Just in case you might not familiar with CJK: https://en.wikipedia.org/wiki/Halfwidth_and_fullwidth_forms

![image](https://github.com/remarkjs/remark-lint/assets/4435445/1c9c0eb8-bde7-46c8-8150-35631f6dd20a)

<img width="652" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/c5211031-19af-4a19-b523-70f4330d28d5">

**Comment by wooorm** [comment]
I’m looking at your code example here on GH and they’re both displayed “misaligned”

**Comment by simnalamburt** [comment]
I don't think it's a good idea to ignore this issue to satisfy GitHub's misrepresentation of CJK characters.

**Comment by simnalamburt** [comment]
I at least want to solve this problem by creating a linter option that respects the size of full-width characters. Since it's an option, it shouldn't affect existing behavior. I strongly believe that this should be the default behavior, not an option to opt-in, and this is what any CJK developer would have done. But if you disagree, I'd be happy to see it fixed as an option. What do you think?

**Comment by wooorm** [comment]
GH is using the web. CSS. A monospaced font. Many editors use the web/css/monospace.

My point is that all tools are going to display it differently.
What is the point in having an option to this rule which will look fine for some users and broken for other users.

That's what the rule currently does: broken for some, good for some. So I suggest not using this rule.

You say that all tools behave one way. I show GH doing it differently. I know editors display emoji differently. Can you provide screenshots for different tools displaying your examples correctly?

**Comment by simnalamburt** [comment]
This it it

<img width="400" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/c5211031-19af-4a19-b523-70f4330d28d5">

Emoji containing version if you need:

<img width="200" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/3b52e6d0-c60c-40e2-9aeb-cbdcd7636bf7">

**Comment by wooorm** [comment]
Everyone uses emoji. Not just Koreans.

> > different tools

Can you please show different editors? And label which editors you show?

**Comment by simnalamburt** [comment]
neovim:
<img width="400" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/7b0b78e5-06f0-4800-9c10-73ca70c60f61">

iTerm2:
<img width="400" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/35dc68eb-c8f4-4a76-93c9-571d609c30a2">

VSCode:
<img width="400" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/d0bec471-fe51-4356-af21-4dfdaa14a904">

zed:
<img width="400" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/3f8bafe6-03c0-4f52-9432-2c9f6ad8869f">

I'll upload windows examples tomorrow since I cannot access my windows devices for right now.

**Comment by xnuk** [comment]
> GH is using the web. CSS. A monospaced font. Many editors use the web/css/monospace.

And many editors try to vertically align a CJK character with 2 latin alphabets, regardless of font configs. Terminal-based editors already do this - every Unicode-aware terminal emulators try to treat a CJK character width as 2 latin chars by default.

[Ace, a web-based editor](https://ace.c9.io/) (which is used by [Rust Playground](https://play.rust-lang.org/)), tries to vertically align intentionally. They measures font size and manually put twice size of it, like: `<span style="width: 14.3984px;" class="ace_cjk">한</span>`.

![image](https://github.com/remarkjs/remark-lint/assets/3071003/51c02559-b95c-4e75-8e6d-74785533ac83)

GitHub's align isn't right because, in my computer, they don't use *a* monospaced font, but they use *two* fonts: 'Liberation Mono' and 'Noto Sans Mono CJK KR' (because Liberation Mono doesn't have any CJK chars). Using only 'Noto Sans Mono CJK KR' mitigates the problem:

![image](https://github.com/remarkjs/remark-lint/assets/3071003/b0968e55-f6b3-4b8b-8a89-96298fab9b35)

Using popular alternative fonts like Neo둥근모 or D2Coding just fits right:

| Neo둥근모 | D2Coding |
|---|---|
| ![image](https://github.com/remarkjs/remark-lint/assets/3071003/0ad1e974-2aa5-4123-a70f-45f44e3eb41d) | ![image](https://github.com/remarkjs/remark-lint/assets/3071003/cfc5c3d7-fe2c-4465-bd93-005c36e1529e) |

**Comment by xnuk** [comment]
Yes, character width can be different by editors, and by fonts. But the point is, you should Unicode-aware if you do "align" something, and 'treat CJK character width as same as two latin alphabets' is a good practice and sensible one. [`wcwidth`](https://man.archlinux.org/man/wcwidth.3.en)-like functions can help this.

**Comment by simnalamburt** [comment]
Similar issue found at 'remark-lint-table-cell-padding'. While code below is compact enough, remark-lint-table-cell-padding reports linter errors.

<img width="402" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/78bb6812-e5d3-426c-8f44-b9d87da8475e">


Plaintext:

```markdown
|Enum Value  |Description   |
|------------|--------------|
|ORDER       |주문 정산     |
|ORDER_CANCEL|주문 취소 정산|
|MANUAL      |수동 정산     |
```

**Comment by simnalamburt** [comment]
Similar issue found at 'remark-lint-maximum-line-length'. While code below is obviously crossed 80-width column, remark-lint-maximum-line-length does not report error.

<img width="937" alt="image" src="https://github.com/remarkjs/remark-lint/assets/4435445/ef8df41f-2c5b-4c06-b51d-37dd29808c1f">


Plaintext:

```
Mercury mercury mercury mercury mercury mercury mercury mercury mercury mercury
mercury.

한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~ 한국어~
```

**Comment by wooorm** [comment]
> GitHub's align isn't right because, in my computer, they don't use a monospaced font, but they use two fonts: 'Liberation Mono' and 'Noto Sans Mono CJK KR' (because Liberation Mono doesn't have any CJK chars). Using only 'Noto Sans Mono CJK KR' mitigates the problem:

Every tool (that is reasonably good) does that, your editor too.
There are many Unicode characters. Fonts don’t have glyphs for all Unicode characters.
So they use a “font stack”. A preferred font renders a glyph, if not found, then the next font is used, and so forth.

> Using popular alternative fonts like Neo둥근모 or D2Coding just fits right:

Right, so this comes back to my original point. How things show, depends on the final user’s computer, and what that user chooses.
Of course, CJK users have configured their editor to display CJK well.
Then, some other user comes in, who uses GitHub, and there the alignment is different.

Different users will see things differently. It is not known to `remark-lint` how things will display.

> But the point is, you should Unicode-aware if you do "align" something, and 'treat CJK character width as same as two latin alphabets' is a good practice and sensible one. wcwidth-like functions can help this.

That is a strong statements while the topic remains vague.

What is “Unicode-aware”? Please **explain or link to a specification or algorithm**.
What about Emoji?
What about ANSI colors?
Control characters?
What do you want to do about [ambiguous characters](https://www.unicode.org/reports/tr11/#Ambiguous)?
Which Unicode version do you want to go with?

Why should `remark-lint` align differently than GitHub?

`wcwidth` is indeed interesting. There is also https://www.npmjs.com/package/wcwidth, which is a JavaScript version, but not maintained. I have also used [`string-width`](https://github.com/sindresorhus/string-width) in the past.

We do support configurable functions in `remark-gfm`: https://github.com/remarkjs/remark-gfm#example-stringlength.
I am open to such functionality of course.
But it needs to work well.

And, could there be a good default that works for every user?

**Comment by simnalamburt** [comment]
> Please explain or link to a specification or algorithm.

OK.

### Specification
- http://www.unicode.org/reports/tr11/

### Algorithm
In order of decreasing precedence:

1. In the following cases, the width of a string differs from the sum of the widths of its constituent characters:
   - The sequence `"\r\n"` has width 1.
   - Emoji-specific ligatures:
     - Well-formed, fully-qualified [emoji ZWJ sequences] have width 2.
     - [Emoji modifier sequences] have width 2.
     - [Emoji presentation sequences] have width 2.
     - Outside of an East Asian context, [text presentation sequences] have width 1 if their base character:
       - Has the [`Emoji_Presentation`] property, and
       - Is not in the [Enclosed Ideographic Supplement] block.
   - Script-specific ligatures:
     - For all the following ligatures, the insertion of any number of [default-ignorable][`Default_Ignorable_Code_Point`] [combining marks] anywhere in the sequence will not change the total width. In addition, for all non-Arabic ligatures, the insertion of any number of [`'\u{200D}'` ZERO WIDTH JOINER](https://www.unicode.org/versions/Unicode15.0.0/ch23.pdf#G23126)s will not affect the width.
     - **[Arabic]**: A character sequence consisting of one character with [`Joining_Group`]`=Lam`, followed by any number of characters with [`Joining_Type`]`=Transparent`, followed by one character with [`Joining_Group`]`=Alef`, has total width 1. For example: `لا`‎, `لآ`‎, `ڸا`‎, `لٟٞأ`
     - **[Buginese]**: `"\u{1A15}\u{1A17}\u{200D}\u{1A10}"` (<a, -i> ya, `ᨕᨗ‍ᨐ`) has total width 1.
     - **[Hebrew]**: `"א\u{200D}ל"` (Alef-Lamed, `א‍ל`) has total width 1.
     - **[Khmer]**: Coeng signs consisting of `'\u{17D2}'` followed by a character in `'\u{1780}'..='\u{1782}' | '\u{1784}'..='\u{1787}' | '\u{1789}'..='\u{178C}' | '\u{178E}'..='\u{1793}' | '\u{1795}'..='\u{1798}' | '\u{179B}'..='\u{179D}' | '\u{17A0}' | '\u{17A2}'  | '\u{17A7}' | '\u{17AB}'..='\u{17AC}' | '\u{17AF}'` have width 0.
     - **[Lisu]**: Tone letter combinations consisting of a character in the range `'\u{A4F8}'..='\u{A4FB}'` followed by a character in the range `'\u{A4FC}'..='\u{A4FD}'` have width 1. For example: `ꓹꓼ`
     - **[Old Turkic]**: `"\u{10C32}\u{200D}\u{10C03}"` (`𐰲‍𐰃`) has total width 1.
     - **[Tifinagh]**: A sequence of a Tifinagh consonant in the range `'\u{2D31}'..='\u{2D65}' | '\u{2D6F}'`, followed by either [`'\u{2D7F}'` TIFINAGH CONSONANT JOINER] or `'\u{200D}'`, followed by another Tifinangh consonant, has total width 1. For example: `ⵏ⵿ⴾ`
   - In an East Asian context only, `<`, `=`, or `>` have width 2 when followed by [`'\u{0338}'` COMBINING LONG SOLIDUS OVERLAY]. The two characters may be separated by any number of characters whose canonical decompositions consist only of characters meeting one of the following requirements:
     - Has [`Canonical_Combining_Class`] greater than 1, or
     - Is a [default-ignorable][`Default_Ignorable_Code_Point`] [combining mark][combining marks].
2. In all other cases, the width of the string equals the sum of its character widths:
   1. [`'\u{2D7F}'` TIFINAGH CONSONANT JOINER] has width 1 (outside of the ligatures described previously).
   2. [`'\u{115F}'` HANGUL CHOSEONG FILLER](https://util.unicode.org/UnicodeJsps/character.jsp?a=115F) and [`'\u{17A4}'` KHMER INDEPENDENT VOWEL QAA](https://util.unicode.org/UnicodeJsps/character.jsp?a=17A4) have width 2.
   3. [`'\u{17D8}'` KHMER SIGN BEYYAL](https://util.unicode.org/UnicodeJsps/character.jsp?a=17D8) has width 3.
   4. The following have width 0:
      - [Characters](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5Cp%7BDefault_Ignorable_Code_Point%7D) with the [`Default_Ignorable_Code_Point`] property.
      - [Characters](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5Cp%7BGrapheme_Extend%7D) with the [`Grapheme_Extend`] property.
      - The following 8 characters, all of which have NFD decompositions consisting of two [`Grapheme_Extend`] characters:
        - [`'\u{0CC0}'` KANNADA VOWEL SIGN II](https://util.unicode.org/UnicodeJsps/character.jsp?a=0CC0),
        - [`'\u{0CC7}'` KANNADA VOWEL SIGN EE](https://util.unicode.org/UnicodeJsps/character.jsp?a=0CC7),
        - [`'\u{0CC8}'` KANNADA VOWEL SIGN AI](https://util.unicode.org/UnicodeJsps/character.jsp?a=0CC8),
        - [`'\u{0CCA}'` KANNADA VOWEL SIGN O](https://util.unicode.org/UnicodeJsps/character.jsp?a=0CCA),
        - [`'\u{0CCB}'` KANNADA VOWEL SIGN OO](https://util.unicode.org/UnicodeJsps/character.jsp?a=0CCB),
        - [`'\u{1B3B}'` BALINESE VOWEL SIGN RA REPA TEDUNG](https://util.unicode.org/UnicodeJsps/character.jsp?a=1B3B),
        - [`'\u{1B3D}'` BALINESE VOWEL SIGN LA LENGA TEDUNG](https://util.unicode.org/UnicodeJsps/character.jsp?a=1B3D), and
        - [`'\u{1B43}'` BALINESE VOWEL SIGN PEPET TEDUNG](https://util.unicode.org/UnicodeJsps/character.jsp?a=1B43).
      - [Characters](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5Cp%7BHangul_Syllable_Type%3DV%7D%5Cp%7BHangul_Syllable_Type%3DT%7D) with a [`Hangul_Syllable_Type`] of `Vowel_Jamo` (`V`) or `Trailing_Jamo` (`T`).
      - The following [`Prepended_Concatenation_Mark`]s:
        - [`'\u{0605}'` NUMBER MARK ABOVE](https://util.unicode.org/UnicodeJsps/character.jsp?a=0605),
        - [`'\u{070F}'` SYRIAC ABBREVIATION MARK](https://util.unicode.org/UnicodeJsps/character.jsp?a=070F),
        - [`'\u{0890}'` POUND MARK ABOVE](https://util.unicode.org/UnicodeJsps/character.jsp?a=0890),
        - [`'\u{0891}'` PIASTRE MARK ABOVE](https://util.unicode.org/UnicodeJsps/character.jsp?a=0891), and
        - [`'\u{08E2}'` DISPUTED END OF AYAH](https://util.unicode.org/UnicodeJsps/character.jsp?a=08E2).
      - [Characters](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5Cp%7BGrapheme_Cluster_Break%3DPrepend%7D-%5Cp%7BPrepended_Concatenation_Mark%7D) with the [`Grapheme_Extend=Prepend`] property, that are not also [`Prepended_Concatenation_Mark`]s.
      - [`'\u{A8FA}'` DEVANAGARI CARET](https://util.unicode.org/UnicodeJsps/character.jsp?a=A8FA).
   5. [Characters](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5Cp%7BEast_Asian_Width%3DF%7D%5Cp%7BEast_Asian_Width%3DW%7D) with an [`East_Asian_Width`] of [`Fullwidth`] or [`Wide`] have width 2.
   6. Characters fulfilling all of the following conditions have width 2 in an East Asian context, and width 1 otherwise:
      - Has an [`East_Asian_Width`] of [`Ambiguous`], or has a canonical decomposition to an [`Ambiguous`] character followed by [`'\u{0338}'` COMBINING LONG SOLIDUS OVERLAY], or is [`'\u{0387}'` GREEK ANO TELEIA](https://util.unicode.org/UnicodeJsps/character.jsp?a=0387), and
      - Does not have a [`General_Category`] of `Letter` or `Modifier_Symbol`.
   7. All other characters have width 1.

[`'\u{0338}'` COMBINING LONG SOLIDUS OVERLAY]: https://util.unicode.org/UnicodeJsps/character.jsp?a=0338
[`'\u{2D7F}'` TIFINAGH CONSONANT JOINER]: https://util.unicode.org/UnicodeJsps/character.jsp?a=2D7F

[`Canonical_Combining_Class`]: https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G50313
[`Default_Ignorable_Code_Point`]: https://www.unicode.org/versions/Unicode15.0.0/ch05.pdf#G40095
[`East_Asian_Width`]: https://www.unicode.org/reports/tr11/#ED1
[`Emoji_Presentation`]: https://unicode.org/reports/tr51/#def_emoji_presentation
[`General_Category`]: https://www.unicode.org/versions/Unicode15.0.0/ch04.pdf#G124142
[`Grapheme_Extend=Prepend`]: https://www.unicode.org/reports/tr29/#Prepend
[`Grapheme_Extend`]: https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G52443
[`Hangul_Syllable_Type`]: https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G45593
[`Joining_Group`]: https://www.unicode.org/versions/Unicode14.0.0/ch09.pdf#G36862
[`Joining_Type`]: http://www.unicode.org/versions/Unicode15.0.0/ch09.pdf#G50009
[`Prepended_Concatenation_Mark`]: https://www.unicode.org/versions/Unicode15.0.0/ch23.pdf#G37908
[`Script`]: https://www.unicode.org/reports/tr24/#Script

[`Fullwidth`]: https://www.unicode.org/reports/tr11/#ED2
[`Wide`]: https://www.unicode.org/reports/tr11/#ED4
[`Ambiguous`]: https://www.unicode.org/reports/tr11/#ED6

[combining marks]: https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G30602

[emoji ZWJ sequences]: https://www.unicode.org/reports/tr51/#def_emoji_sequence
[Emoji modifier sequences]: https://www.unicode.org/reports/tr51/#def_emoji_modifier_sequence
[Emoji presentation sequences]: https://unicode.org/reports/tr51/#def_emoji_presentation_sequence
[text presentation sequences]: https://unicode.org/reports/tr51/#def_text_presentation_sequence

[Enclosed Ideographic Supplement]: https://unicode.org/charts/nameslist/n_1F200.html

[Arabic]: https://www.unicode.org/versions/Unicode15.0.0/ch09.pdf#G7480
[Buginese]: https://www.unicode.org/versions/Unicode15.0.0/ch17.pdf#G26743
[Hebrew]: https://www.unicode.org/versions/Unicode15.0.0/ch09.pdf#G6528
[Khmer]: https://www.unicode.org/versions/Unicode15.0.0/ch16.pdf#G64642
[Lisu]: https://www.unicode.org/versions/Unicode15.0.0/ch18.pdf#G44587
[Old Turkic]: https://www.unicode.org/versions/Unicode15.0.0/ch14.pdf#G41975
[Tifinagh]: http://www.unicode.org/versions/Unicode15.0.0/ch19.pdf#G43184

**Comment by wooorm** [comment]
That‘s 1 of the 30 questions I asked. What about the others?

**Comment by simnalamburt** [comment]
The discussions in this issue tracker are becoming increasingly frustrating.

I mentioned that I had already created a patch because I thought you might find it cumbersome to fix this issue: https://github.com/remarkjs/remark-lint/commit/1e659c27c530c345f2a2c592b572ee15dd5ba62c.

I believe I have sufficiently demonstrated that full-width and half-width characters are standardized concepts in Unicode and how they should operate. I think I have clearly explained in my response on September 27 how to handle Emojis, ANSI colors, Control characters, and ambiguous characters, as well as which Unicode version to follow.

I believe that correctly handling full-width characters is a good default, but I have already stated on March that I agree not to change the default because doing so would break backward compatibility.

remark-gfm already has the functionality to customize `stringLength` with a function of your choice: https://github.com/remarkjs/remark-gfm#example-stringlength. I simply want the same option to be available in remark-lint. (This way, the output generated by remark-gfm will automatically pass the checks by remark-lint.) By doing so, existing users who are not using the `stringLength` option will not experience any changes in behavior. Defining `stringLength` becomes the responsibility of the user rather than the library, so the maintainers of remark-lint do not need to follow-up the complex Unicode standards.

Please provide me a rough guide on how to add the `stringLength` option to remark-lint. Then I will create a patch.

**Comment by finalchild** [comment]
> That‘s 1 of the 30 questions I asked. What about the others?

Maybe you should read the explanation someone is kindly providing to you before responding?

**Comment by simnalamburt** [comment]
Thanks for the patches!
